File: src/entities/instances/model/instance.ts
import type { ModLoader } from '@/entities/minecrafts/@x/instances';
import type { ContentType } from './contentType';
import type {
  Hooks,
  MemorySettings,
  WindowSize,
} from '@/entities/settings/@x/instances';

export interface InstancePluginSettings {
  preLaunch?: string;
}

export interface PackInfo {
  packType: string;
  version: string;
  canUpdate: boolean;
}

export interface Instance {
  id: string;
  path: string;

  name: string;
  iconPath?: string;

  installStage: InstanceInstallStage;

  // Main minecraft metadata
  gameVersion: string;
  loader: ModLoader;
  loaderVersion?: string;

  // Launch arguments
  javaPath?: string;
  extraLaunchArgs?: string[];
  customEnvVars?: Array<[string, string]>;

  // Minecraft runtime settings
  memory?: MemorySettings;
  forceFullscreen?: boolean;
  gameResolution?: WindowSize;

  // Additional information
  created: Date;
  modified: Date;
  lastPlayed?: string;

  timePlayed: number;
  recent_time_played: number;

  hooks: Hooks;

  packInfo?: PackInfo;
}

export interface InstanceFile {
  hash: string;
  name?: string;
  fileName: string;
  size: number;
  contentType: ContentType;
  path: string;
  disabled: boolean;
  update: Record<string, Record<string, unknown>>;
}

export enum InstanceInstallStage {
  /// Instance is installed
  Installed = 'installed',
  /// Instance's minecraft game is still installing
  Installing = 'installing',
  /// Instance created for pack, but the pack hasn't been fully installed yet
  PackInstalling = 'pack_installing',
  /// Instance is not installed
  NotInstalled = 'not_installed',
}

export interface NewInstance {
  name: string;
  gameVersion: string;
  modLoader: ModLoader;
  loaderVersion?: string;
  iconPath?: string;
  skipInstallProfile?: boolean;
  packInfo?: PackInfo;
}

export interface EditInstance {
  name?: string;
  javaPath?: string | null;
  extraLaunchArgs?: string[] | null;
  customEnvVars?: Array<[string, string]> | null;
  memory?: MemorySettings | null;
  gameResolution?: WindowSize | null;
}

export interface InstanceImportDto {
  packType: string;
  path: string;
}

export interface ImportHandler {
  packType: string;
  title: string;
  fieldLabel: string;
  fileName: string;
  fileExtensions: string[];
}

File: packages/core/aether-core/src/features/plugins/infra/extism/mappers/instance/instance_.rs
use aether_core_plugin_api::v0::InstanceDto;

use crate::features::instance::Instance;

impl From<Instance> for InstanceDto {
    fn from(value: Instance) -> Self {
        Self {
            id: value.id().to_owned(),

            name: value.name().to_owned(),

            icon_path: value.icon_path().map(ToString::to_string),

            install_stage: value.install_stage.into(),

            game_version: value.game_version,
            loader: value.loader.into(),
            loader_version: value.loader_version.map(Into::into),

            java_path: value.java_path.into(),
            launch_args: value.launch_args.into(),
            env_vars: value.env_vars.into(),

            memory: value.memory.into(),

            window: value.window.into(),

            created: value.created,
            modified: value.modified,
            last_played: value.last_played,

            time_played: value.time_played,
            recent_time_played: value.recent_time_played,

            hooks: value.hooks.into(),

            pack_info: value.pack_info.map(Into::into),
        }
    }
}
